core/src/modules/configWriter.ts

```diff
@@ -36,15 +64,78 @@ export namespace ConfigWriter {
     return configObjects;
   }
 
-  export async function setFileLoaded(absFilePath: string) {
-    FILES_TO_DELETE.push(absFilePath);
+  export async function getConfigObjects(): Promise<WritableConfigObject[]> {
+    let objects = [];
+    const queryParams = { where: { locked: null } };
+    const queries = {
+      apps: await App.findAll(queryParams),
+      sources: await Source.findAll(queryParams),
+      schedules: await Schedule.findAll(queryParams),
+      properties: await Property.findAll(queryParams),
+      groups: await Group.findAll(queryParams),
+      destinations: await Destination.findAll(queryParams),
+    };
+
+    for (let [type, instances] of Object.entries(queries)) {
+      for (let instance of instances) {
+        const object = await instance.getConfigObject();
+        const filePath = `${type}/${instance.id}.json`;
+        objects.push({ filePath, object });
+      }
+    }
+    return objects;
+  }
+
+  // ---------------------------------------- | Config File Cache
+
+  export function getConfigFileCache() {
+    return CONFIG_FILE_CACHE;
+  }
+
+  export function resetConfigFileCache() {
+    CONFIG_FILE_CACHE = [];
   }
 
+  export function getLockKey(configObject: ConfigurationObject): string | null {
+    if (process.env.GROUPAROO_RUN_MODE !== "cli:config") {
+      return getCodeConfigLockKey();
+    }
+    if (isLockable(configObject)) {
+      return "config:writer";
+    }
+    return null;
+  }
+
+  function fileIsLockable(absFilePath: string): boolean {
+    // Otherwise, it is lockable if it is a JS file.
+    const ext = path.extname(absFilePath);
+    return ext === ".js";
+  }
+
+  function isLockable(object: ConfigurationObject) {
+    const isMatch = (o) => o.id === object.id && o.class === object.class;
+    const cachedFileObj: CachedConfigFile = CONFIG_FILE_CACHE.find(
+      (cache) => cache.objects.filter(isMatch).length > 0
+    );
```

Comment:
Will this get weird when https://github.com/grouparoo/grouparoo/pull/1795 is merged in (2 objects in one file)?

Reply:
It _shouldn't_. This checks against the object itself. So if both a source and a schedule are loaded from the same file, they will both either be locked or not, based on the file extension.

We don't have to worry about the object being an array based on how this is being used today.